libdebug/cffi/ptrace_cffi_source.c

```diff
@@ -67,10 +78,310 @@ struct thread_status {
+}
+
+
+void remove_hardware_breakpoint(struct hardware_breakpoint *bp)
+{
+    // find the register
+    int i;
+    for (i = 0; i < 4; i++) {
+        unsigned long address = ptrace(PTRACE_PEEKUSER, bp->tid, DR_BASE + i * DR_SIZE);
+
+        if (address == bp->addr)
+            break;
+    }
+
+    if (i == 4) {
+        perror("Breakpoint not found");
+        return;
+    }
+
+    // read the state from DR7
+    unsigned long state = ptrace(PTRACE_PEEKUSER, bp->tid, DR_BASE + 7 * DR_SIZE);
+
+    // reset the state
+    state &= ~(3 << CTRL_COND(i));
+    state &= ~(3 << CTRL_LEN(i));
+
+    // write the state
+    ptrace(PTRACE_POKEUSER, bp->tid, DR_BASE + 7 * DR_SIZE, state);
+
+    // clear the address
+    ptrace(PTRACE_POKEUSER, bp->tid, DR_BASE + i * DR_SIZE, 0);
+}
+
+int is_breakpoint_hit(struct hardware_breakpoint *bp)
+{
+    unsigned long status = ptrace(PTRACE_PEEKUSER, bp->tid, DR_BASE + 6 * DR_SIZE);
+
+    int index;
+    if (status & 0x1)
+        index = 0;
+    else if (status & 0x2)
+        index = 1;
+    else if (status & 0x4)
+        index = 2;
+    else if (status & 0x8)
+        index = 3;
+    else
+        return 0;
+
+    unsigned long address = ptrace(PTRACE_PEEKUSER, bp->tid, DR_BASE + index * DR_SIZE);
+
+    if (address == bp->addr)
+        return 1;
+
+    return 0;
+}
+
+int get_remaining_hw_breakpoint_count(struct global_state *state, int tid)
+{
+    int i;
+    for (i = 0; i < 4; i++) {
+        unsigned long address = ptrace(PTRACE_PEEKUSER, tid, DR_BASE + i * DR_SIZE);
+
+        if (!address)
+            break;
+    }
+
+    return 4 - i;
+}
+
+int get_remaining_hw_watchpoint_count(struct global_state *state, int tid)
+{
+    return get_remaining_hw_breakpoint_count(state, tid);
+}
+#endif
+
+#ifdef ARCH_AARCH64
+struct user_hwdebug_state {
+    unsigned int dbg_info;
+	unsigned int pad;
+	struct {
+		unsigned long addr;
+		unsigned int ctrl;
+		unsigned int pad;
+	} dbg_regs[16];
+};
+
+int get_breakpoint_type(char type[2])
+{
+    if (type[0] == 'r') {
+        if (type[1] == 'w') {
+            return 3;
+        } else {
+            return 1;
+        }
+    } else if (type[0] == 'w') {
+        return 2;
+    } else if (type[0] == 'x') {
+        return 0;
+    } else {
+        return -1;
+    }
+}
+
+void install_hardware_breakpoint(struct hardware_breakpoint *bp)
+{
+    // find a free debug register
+    struct user_hwdebug_state state = {0};
+
+    struct iovec iov;
+    iov.iov_base = &state;
+    iov.iov_len = sizeof state;
+
+    unsigned long command = get_breakpoint_type(bp->type) == 0 ? NT_ARM_HW_BREAK : NT_ARM_HW_WATCH;
+
+    ptrace(PTRACE_GETREGSET, bp->tid, command, &iov);
+
+    int i;
+    for (i = 0; i < 16; i++) {
+        if (!state.dbg_regs[i].addr)
+            break;
+    }
+
+    if (i == 16) {
+        perror("No debug registers available");
+        return;
+    }
+
+    if (bp->type[0] == 'x') {
+        // Hardware breakpoint can only be of length 4
```

Comment:
This is not enforced in the python `validate_breakpoint_aarch64` function.

Reply:
The pydoc for the breakpoint function says that length is supported only for watchpoints

Reply:
So we just don't bother validating the length, the CFFI sets it to 4